packages/rest/src/lib/handlers/SequentialHandler.ts

```diff
@@ -90,23 +123,56 @@ export class SequentialHandler {
 		// Wait for any previous requests to be completed before this one is run
 		await this.#asyncQueue.wait();
 		try {
-			// Wait for any global rate limits to pass before continuing to process requests
-			await this.manager.globalTimeout;
-			// Check if this request handler is currently rate limited
-			if (this.limited) {
+			/*
+			 * After calculations have been done, pre-emptively stop further requests
+			 * Potentially loop until this task can run if e.g. the global rate limit is hit twice
+			 */
+			while (this.limited) {
+				const isGlobal = this.globalLimited;
+				let limit: number;
+				let timeout: number;
+				let delay: Promise<void>;
+
+				if (isGlobal) {
+					// Set RateLimitData based on the globl limit
+					limit = this.manager.options.globalLimit;
+					timeout = this.manager.globalReset + this.manager.options.offset - Date.now();
+					// If this is the first task to reach the global timeout, set the global delay
+					if (!this.manager.globalDelay) {
+						// The global delay function clears the global delay state when it is resolved
+						this.manager.globalDelay = this.globalDelayFor(timeout);
+					}
+					delay = this.manager.globalDelay;
+				} else {
+					// Set RateLimitData based on the route-specific limit
+					limit = this.limit;
+					timeout = this.timeToReset;
+					delay = sleep(timeout);
```

Comment:
unref?

Reply:
There's a lot of places that aren't unreffed right now, I was going to do them in a separate PR, but I'll do all the ones I added at least.